### PR TITLE
os/filestore: avoid unnecessary copy in filestore::_do_transaction

### PR DESCRIPTION
**Pull request by wwformat**

### CODE REVIEW

File: src/os/filestore/FileStore.cc [hunk 1]
```diff
@@ -2530,9 +2530,10 @@ void FileStore::_do_transaction(
       break;
     case Transaction::OP_TOUCH:
       {
-        coll_t cid = i.get_cid(op->cid);
+        const coll_t &_cid = i.get_cid(op->cid);
         const ghobject_t &oid = i.get_oid(op->oid);
-	_kludge_temp_object_collection(cid, oid);
+        const coll_t &cid = !_need_temp_object_collection(_cid, oid) ?
+          _cid : _cid.get_temp();
```
Comment (liewegas) [inline src/os/filestore/FileStore.cc]: does this work to assign a const coll_t& ref to a temporary return value?

Comment (wwformat) [inline src/os/filestore/FileStore.cc]: @liewegas 
Yes, I think const XXX & can reference a temporary object, and XXX & cannot.
And I found a similar change in commit 18fe93c67762ef70cd88609e253ad65e2a832529, like this:
@@ -2882,10 +2882,10 @@ void FileStore::_do_transaction(
 // --------------------
 // objects

-bool FileStore::exists(coll_t cid, const ghobject_t& oid)
+bool FileStore::exists(const coll_t& _cid, const ghobject_t& oid)
 {
-  tracepoint(objectstore, exists_enter, cid.c_str());
-  _kludge_temp_object_collection(cid, oid);
+  tracepoint(objectstore, exists_enter, _cid.c_str());
+  const coll_t& cid = !_need_temp_object_collection(_cid, oid) ? _cid : _cid.get_temp();
   struct stat st;
   bool retval = stat(cid, oid, &st) == 0;
   tracepoint(objectstore, exists_exit, retval);
t_collection(cid, oid);
+  tracepoint(objectstore, exists_enter, _cid.c_str());
+  const coll_t& cid = !_need_temp_object_collection(_cid, oid) ? _cid : _cid.get_temp();

**Comment by wwformat** [comment]
@liewegas ping

File: src/os/filestore/FileStore.cc (hunk 1 above)
Comment (tchaikov) [inline src/os/filestore/FileStore.cc]: yes, it's valid. see http://stackoverflow.com/questions/2784262/does-a-const-reference-prolong-the-life-of-a-temporary

**Comment by tchaikov** [comment]
@liewegas i think this change makes sense to me albeit more verbose than before due to the limitation of C++ syntax.

**Comment by liewegas** [comment]
Yeah; let's test!  I'm a little nervous just because it's many opportunities to introduce a bug via a typo.  @wwformat is there a measureable performance impact?

**Comment by yuriw** [comment]
@liewegas @tchaikov fyi

This was part of https://trello.com/c/6ZL8w8JQ and passed tests

http://pulpito.ceph.com/yuriw-2017-03-06_21:48:56-rados-wip-yuri-testing2_2017_3_7-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-03-07_15:31:16-rados-wip-yuri-testing2_2017_3_7---basic-smithi/

**Pull request merged**